wrappers/wrapper.rb

```diff
@@ -684,70 +602,119 @@ def detect_inconsistent_relation_timewindows_od_service(vrp, unfeasible, service
       }
     end
 
-    def service_reachable_by_vehicle_within_timewindows(vrp, activity, vehicle)
-      vehicle_start = vehicle.timewindow&.start || vehicle.sequence_timewindows.collect(&:start).min || 0
-      vehicle_end =
-        if vehicle.cost_late_multiplier&.positive? # vehicle lateness is allowed
-          vehicle.timewindow&.end ?
-            vehicle.timewindow.end + vehicle.timewindow.maximum_lateness :
-            vehicle.sequence_timewindows.collect{ |tw| tw.end + tw.maximum_lateness }.max
-        else # vehicle lateness is not allowed
-          vehicle.timewindow&.end || vehicle.sequence_timewindows.collect(&:end).max
-        end
+    def no_vehicle_with_compatible_skills(vrp, service)
+      service.vehicle_compatibility ||= {}
 
-      matrix = vrp.matrices.find{ |m| m.id == vehicle.matrix_id }
+      vrp.vehicles.each{ |vehicle|
+        service.vehicle_compatibility[vehicle.id] ||=
+          service.vehicle_compatibility[vehicle.id].nil? && # if it is true or false, no need to recheck
+          (service.skills.empty? || vehicle.skills.any?{ |v_skill_set| (service.skills - v_skill_set).empty? }) &&
+          (service.sticky_vehicles.empty? || service.sticky_vehicle_ids.include?(vehicle.id))
```

Comment:
It looks like there still is `sticky_vehicles` under services as a member :/  I was going to remove the last condition but isn't it better to keep it until `sticky_vehicles` is removed -- because it doesn't really affect the performance because if the vehicle is incompatible (due to sticky) it is detected via the previous condition (through skills) and the last condition is never reached. If it is compatible (with sticky vehicle) the last condition checks if a very shor array includes an item -- `service.sticky_vehicle_ids.include?(vehicle.id)`.

I wanted to keep it thinking sticky_vehicles might behave different internally -- to skills (for example perriodic expansion) but if you are sure that checking sticky is never needed, I can delete it anyways. Let me know, i think it was you who did that dev so you know better.

Reply:
I kept sticky_vehicles and original_skills mostly in case of output purpose. I expect to have the sticky vehicles removed from the internal logic. It makes no computational difference here, but we have an unreachable condition, which doesn't sounds like a good idea.

Reply:
It is reachable if the service has a sticky_vehicle and we are checking for this vehicle. It's like doublly verified :D I would prefer to leave it and let it be removed while we remove the sticky_vehicles but it looks like you are sure that we don't have or won't add extra logic to sticky_vehicles in the mean time, so I will delete it ;)